### PR TITLE
llama : add infill sampler

### PR DESCRIPTION
**Pull request by ggerganov**
Add a new sampler that I think is suitable for infill tasks. It promotes end-of-generation (EOG) tokens if it is not very confident and also combines (merges) tokens with common prefix.

I think there are more improvements that can be made specifically for fill-in-the-middle, but this version seems to work OK for now.

```bash
curl \
    --silent --no-buffer --request POST --url http://127.0.0.1:8012/infill \
    --header "Content-Type: application/json" \
    --data '{"input_suffix": "    return 0;\n}\n", "input_prefix": "#include <cstdio>\n\nint main() {\n    printf(", "prompt": "", "top_k": 40, "top_p": 0.99, "samplers": ["top_k", "top_p", "infill"]}' | jq
```

Also add the following function to the `libllama` API:

```c
    // check if token0 is contained as a prefix in token1
    LLAMA_API bool llama_token_is_prefix(
              const struct llama_model * model,
                           llama_token   token0,
                           llama_token   token1);
```

### CODE REVIEW

File: src/llama-sampling.cpp [hunk 1]
```diff
@@ -1644,6 +1644,207 @@ struct llama_sampler * llama_sampler_init_logit_bias(
...
+                p_sum += cur_p->data[i].p;
+
+                cur_p->data[cur_p->size++] = cur_p->data[i];
+            }
+        }
+
+        // normalize probs
+        for (size_t i = 0; i < cur_p->size; ++i) {
+            cur_p->data[i].p /= p_sum;
+        }
```
Comment (slaren) [inline src/llama-sampling.cpp:1710]: A few samplers do this, but I don't see the point because every sampler that needs the probabilities calls softmax first anyway and recomputes the probabilities.

During the refactor I came to the conclusion that the we only really store logits. Every time probabilities are needed a softmax is done to get them, `llama_token_data::p` is only used as temporary storage for the result of the softmax, and could be removed entirely.

Comment (ggerganov) [reply src/llama-sampling.cpp:1710]: I think currently there is a scenario that uses it - call `dist` sampler without explicit `softmax` before that. We don't do it in any of the examples, but it's technically possible?

Anyway, I agree that the `p` should be removed completely.

Comment (slaren) [reply src/llama-sampling.cpp:1710]: I think that should be considered a bug in the dist sampler then, because there is no way to know if the probabilities are valid without calling softmax. So any sampler that needs them, must call softmax.

Comment (ggerganov) [reply src/llama-sampling.cpp:1710]: Yes, indeed it is a big. There are a few places where we do:

```c++
llama_sampler * smpl = llama_sampler_chain_init(sparams);                                                              
                                                                                                                             
llama_sampler_chain_add(smpl, llama_sampler_init_top_k(params.sparams.top_k));                                      
llama_sampler_chain_add(smpl, llama_sampler_init_top_p(params.sparams.top_p, params.sparams.min_keep));             
llama_sampler_chain_add(smpl, llama_sampler_init_temp (params.sparams.temp));                                                                                                                                                             
llama_sampler_chain_add(smpl, llama_sampler_init_dist (params.sparams.seed));
```

This would render the temperature sampler useless as it modifies only the logits. I think we should remove the explicit `softmax` calls in places like `common/sampling.cpp`:

```c++
llama_sampler_chain_add(result->chain, llama_sampler_init_softmax()); // remove this
llama_sampler_chain_add(result->chain, llama_sampler_init_dist(params.seed));
```

And update the `dist` sampler to do softmax at the start. Sounds good?

Comment (slaren) [reply src/llama-sampling.cpp:1710]: Sounds good. We should probably remove `llama_sampler_init_softmax` entirely since it is useless to applications.
